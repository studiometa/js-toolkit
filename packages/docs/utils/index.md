# Utils

Find below functions that can help you in your projects.

## Misc.

- [debounce](/utils/debounce.html)
- [focusTrap](/utils/focusTrap.html)
- [keyCodes](/utils/keyCodes.html)
- [memoize](/utils/memoize.html)
- [nextFrame](/utils/nextFrame.html)
- [scrollTo](/utils/scrollTo.html)
- [throttle](/utils/throttle.html)

## Collision

- [boundingRectToCircle](/utils/collision/boundingRectToCircle.html)
- [collideCircleCircle](/utils/collision/collideCircleCircle.html)
- [collideCircleRect](/utils/collision/collideCircleRect.html)
- [collidePointCircle](/utils/collision/collidePointCircle.html)
- [collidePointRect](/utils/collision/collidePointRect.html)
- [collideRectRect](/utils/collision/collideRectRect.html)

## CSS

- [matrix](/utils/css/matrix.html)
- [transition](/utils/css/transition.html)

## History

- [historyPush](/utils/history/historyPush.html)
- [historyReplace](/utils/history/historyReplace.html)
- [objectToURLSearchParams](/utils/history/objectToURLSearchParams.html)

## Math

- [clamp](/utils/math/clamp.html)
- [clamp01](/utils/math/clamp01.html)
- [damp](/utils/math/damp.html)
- [inertiaFinalValue](/utils/math/inertiaFinalValue.html)
- [lerp](/utils/math/lerp.html)
- [map](/utils/math/map.html)
- [round](/utils/math/round.html)

## String

- [withLeadingCharacters](/utils/string/withLeadingCharacters.html)
- [withLeadingSlash](/utils/string/withLeadingSlash.html)
- [withoutLeadingCharacters](/utils/string/withoutLeadingCharacters.html)
- [withoutLeadingCharactersRecursive](/utils/string/withoutLeadingCharactersRecursive.html)
- [withoutLeadingSlash](/utils/string/withoutLeadingSlash.html)
- [withoutTrailingCharacters](/utils/string/withoutTrailingCharacters.html)
- [withoutTrailingCharactersRecursive](/utils/string/withoutTrailingCharactersRecursive.html)
- [withoutTrailingSlash](/utils/string/withoutTrailingSlash.html)
- [withTrailingCharacters](/utils/string/withTrailingCharacters.html)
- [withTrailingSlash](/utils/string/withTrailingSlash.html)
