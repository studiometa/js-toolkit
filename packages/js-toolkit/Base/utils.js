import { isArray, isDefined, isDev } from '../utils/index.js';

/**
 * @typedef {import('./index.js').default} Base
 */

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {string} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @returns {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(nameOrSelector, element = document) {
  const selector = `[data-component="${nameOrSelector}"]`;
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}

/**
 * Test if an event is defined in the given config.
 *
 * @param   {string} event
 * @param   {import('./index.js').BaseConfig} config
 * @returns {boolean}
 */
export function eventIsDefinedInConfig(event, config) {
  return isArray(config.emits) && config.emits.includes(event);
}

/**
 * Test if an event can be used on the given element.
 *
 * @param   {string} event
 * @param   {HTMLElement} element
 * @returns {boolean}
 */
export function eventIsNative(event, element) {
  return isDefined(element[`on${event}`]);
}

/**
 * Get the target of a given event.
 *
 * @param   {Base} instance
 * @param   {string} event
 * @param   {import('./index.js').BaseConfig} config
 * @returns {Base|Base['$el']}
 */
export function getEventTarget(instance, event, config) {
  if (eventIsDefinedInConfig(event, config)) {
    return instance;
  }

  if (eventIsNative(event, instance.$el)) {
    return instance.$el;
  }

  if (isDev) {
    console.warn(
      `[${config.name}]`,
      `The "${event}" event is missing from the configuration and is not a native`,
      `event for the root element of type \`${instance.$el.constructor.name}\`.`
    );
  }

  return instance;
}
