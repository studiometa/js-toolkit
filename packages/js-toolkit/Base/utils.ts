import { isArray, isDefined, isDev, SmartQueue } from '../utils/index.js';
import type { Base, BaseConfig } from './index.js';
import { features } from './features.js';

let queue: SmartQueue;

/**
 * Add a task to the main queue.
 */
export function addToQueue(fn: () => unknown): void {
  if (!features.get('asyncChildren')) {
    fn();
    return;
  }

  if (!queue) {
    queue = new SmartQueue();
  }

  queue.add(fn);
}

const selectors = new Map();

// Separator used for multi-component declaration in `data-component` attributeS.
const separator = ' ';

/**
 * Get the selector for a given component.
 */
function getSelector(nameOrSelector: string): string {
  if (!selectors.has(nameOrSelector)) {
    const parts = [
      // Single selector
      `[data-component="${nameOrSelector}"]`,
      // Selector in the middle of a list of selectors
      `[data-component*="${separator}${nameOrSelector}${separator}"]`,
      // Selector at the end of a list of selectors
      `[data-component$="${separator}${nameOrSelector}"]`,
      // Selector at the beginning of a list of selectors
      `[data-component^="${nameOrSelector}${separator}"]`,
    ];
    selectors.set(nameOrSelector, parts.join(','));
  }

  return selectors.get(nameOrSelector);
}

/**
 * Get a list of elements based on the name of a component.
 *
 * @param {string} nameOrSelector
 *   The name or selector to used for this component.
 * @param {HTMLElement|Document} element
 *   The root element on which to query the selector, defaults to `document`.
 * @returns {Array<HTMLElement>}
 *   A list of elements on which the component should be mounted.
 */
export function getComponentElements(
  nameOrSelector: string,
  element: HTMLElement | Document = document,
): HTMLElement[] {
  const selector = getSelector(nameOrSelector);
  let elements = [];

  try {
    elements = Array.from(element.querySelectorAll(selector));
    // eslint-disable-next-line no-empty
  } catch {}

  // If no child component found with the default selector, try a classic DOM selector
  if (elements.length === 0) {
    elements = Array.from(element.querySelectorAll(nameOrSelector));
  }

  return elements;
}

/**
 * Test if an event is defined in the given config.
 */
export function eventIsDefinedInConfig(event: string, config: BaseConfig): boolean {
  return isArray(config.emits) && config.emits.includes(event);
}

/**
 * Test if an event can be used on the given element.
 */
export function eventIsNative(event: string, element: HTMLElement): boolean {
  return isDefined(element[`on${event}`]);
}

/**
 * Get the target of a given event.
 * @todo Return false in v3 if event is not defined or not native to prevent adding a listener.
 */
export function getEventTarget(
  instance: Base,
  event: string,
  config: BaseConfig,
): Base | Base['$el'] {
  if (eventIsDefinedInConfig(event, config)) {
    return instance;
  }

  if (eventIsNative(event, instance.$el)) {
    return instance.$el;
  }

  if (isDev) {
    console.warn(
      `[${config.name}]`,
      `The "${event}" event is missing from the configuration and is not a native`,
      `event for the root element of type \`${instance.$el.constructor.name}\`.`,
    );
  }

  // @todo v3 return false or null
  return instance;
}
