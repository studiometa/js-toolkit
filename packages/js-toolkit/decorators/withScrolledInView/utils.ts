import { isNumber, isString } from '../../utils/index.js';
import type { NormalizedOffset } from './types.js';

const units = {
  px: 'px',
  vh: 'vh',
  vw: 'vw',
  vmin: 'vmin',
  vmax: 'vmax',
  percent: '%',
};

const namedOffsets = {
  start: 0,
  center: 0.5,
  end: 1,
};

/**
 * Normalize offset input.
 */
export function normalizeOffset(offsets: string): NormalizedOffset {
  return offsets
    .split('/')
    .map((offset) => offset.trim().split(' ').slice(0, 2)) as NormalizedOffset;
}

/**
 * Parse the given offset.
 */
export function parseNamedOffset(offset: string | number): string | number {
  if (isNumber(offset)) {
    return offset;
  }

  if (isNumber(namedOffsets[offset])) {
    return namedOffsets[offset];
  }

  if (isNumber(Number(offset))) {
    return Number(offset);
  }

  if (isString(offset) && offset.endsWith(units.percent)) {
    return Number.parseFloat(offset) / 100;
  }

  return offset;
}

/**
 * Get an edge value given a start, size and offset value.
 */
export function getEdgeWithOffset(start: number, size: number, offset: string | number): number {
  if (isNumber(offset)) {
    return start + size * offset;
  }

  const parsedNamedOffset = parseNamedOffset(offset);
  if (isNumber(parsedNamedOffset)) {
    return start + size * parsedNamedOffset;
  }

  if (isString(parsedNamedOffset)) {
    const parsedOffset = Number.parseFloat(parsedNamedOffset);

    if (parsedNamedOffset.endsWith(units.px)) {
      return start + parsedOffset;
    }

    if (parsedNamedOffset.endsWith(units.vh)) {
      return start + (parsedOffset * window.innerHeight) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vw)) {
      return start + (parsedOffset * window.innerWidth) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vmin)) {
      return start + (parsedOffset * Math.min(window.innerWidth, window.innerHeight)) / 100;
    }

    if (parsedNamedOffset.endsWith(units.vmax)) {
      return start + (parsedOffset * Math.max(window.innerWidth, window.innerHeight)) / 100;
    }
  }

  return start;
}

/**
 * Get starting and ending edges for a given axis, a target sizings, a container sizings and their offset.
 */
export function getEdges(axis: 'x' | 'y', targetSizes, containerSizes, offset): [number, number] {
  const sizeKey = axis === 'x' ? 'width' : 'height';
  const targetStart = getEdgeWithOffset(targetSizes[axis], targetSizes[sizeKey], offset[0][0]);
  const containerStart = getEdgeWithOffset(0, containerSizes[sizeKey], offset[0][1]);
  const targetEnd = getEdgeWithOffset(targetSizes[axis], targetSizes[sizeKey], offset[1][0]);
  const containerEnd = getEdgeWithOffset(0, containerSizes[sizeKey], offset[1][1]);
  const start = targetStart - containerStart;
  const end = targetEnd - containerEnd;

  return [start, end];
}
