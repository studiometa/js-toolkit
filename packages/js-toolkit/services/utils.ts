export const ONCE_CAPTURE_EVENT_OPTIONS = { once: true, capture: true };
export const PASSIVE_CAPTURE_EVENT_OPTIONS = { passive: true, capture: true };
export const PASSIVE_EVENT_OPTIONS = { passive: true };
export const CAPTURE_EVENT_OPTIONS = { capture: true };
