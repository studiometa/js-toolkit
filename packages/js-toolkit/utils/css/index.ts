export * from './animate.js';
export { add as addClass, remove as removeClass, toggle as toggleClass } from './classes.js';
export { add as addStyle, remove as removeStyle } from './styles.js';
export { default as getOffsetSizes } from './getOffsetSizes.js';
export { default as matrix } from './matrix.js';
export { default as transform } from './transform.js';
export { default as transition } from './transition.js';
