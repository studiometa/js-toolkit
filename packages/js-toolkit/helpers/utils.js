/* eslint-disable import/prefer-default-export */
import { getComponentElements } from '../Base/utils.js';

/**
 * Get the target elements for the lazy import helper functions.
 *
 * @param {string|HTMLElement|HTMLElement[]} nameOrSelectorOrElement
 *   The original selector or element, or list of elements.
 * @param {HTMLElement} [context]
 *   The optional context to use to query for elements.
 * @returns {HTMLElement[]} A normalized list of elements.
 */
export function getTargetElements(nameOrSelectorOrElement, context) {
  if (typeof nameOrSelectorOrElement === 'string') {
    return getComponentElements(nameOrSelectorOrElement, context);
  }

  if (!Array.isArray(nameOrSelectorOrElement)) {
    return [nameOrSelectorOrElement];
  }

  return nameOrSelectorOrElement;
}
