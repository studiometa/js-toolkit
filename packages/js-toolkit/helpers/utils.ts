/* eslint-disable import/prefer-default-export */
import { getComponentElements } from '../Base/utils.js';
import { isString, isArray } from '../utils/index.js';

/**
 * Get the target elements for the lazy import helper functions.
 *
 * @param {string|HTMLElement|HTMLElement[]} nameOrSelectorOrElement
 *   The original selector or element, or list of elements.
 * @param {HTMLElement} [context]
 *   The optional context to use to query for elements.
 * @return {HTMLElement[]} A normalized list of elements.
 */
export function getTargetElements(
  nameOrSelectorOrElement: string | HTMLElement | HTMLElement[],
  context: HTMLElement,
): HTMLElement[] {
  if (isString(nameOrSelectorOrElement)) {
    return getComponentElements(nameOrSelectorOrElement, context);
  }

  if (!isArray(nameOrSelectorOrElement)) {
    return [nameOrSelectorOrElement];
  }

  return nameOrSelectorOrElement;
}
