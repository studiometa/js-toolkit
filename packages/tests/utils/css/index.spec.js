import * as css from '@studiometa/js-toolkit/utils/css';

describe('@studiometa/js-toolkit/utils/css exports', () => {
  it('should export all scripts', () => {
    expect(Object.keys(css)).toMatchSnapshot();
  });
});
