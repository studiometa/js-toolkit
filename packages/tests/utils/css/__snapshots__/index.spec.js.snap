// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`@studiometa/js-toolkit/utils/css exports should export all scripts 1`] = `
Array [
  "addClass",
  "addStyle",
  "getOffsetSizes",
  "matrix",
  "removeClass",
  "removeStyle",
  "toggleClass",
  "transition",
]
`;
